### DISCUSSION TITLE
Have beatmap/skin/score import not reimport everything that's already there.

### DISCUSSION THREAD

**Discussion by Backoffson** [ppy/osu#19024] [Q&A] [answered]
Edit: This is apparently already a feature and what I'm experiencing is more like a bug or something, so I've changed the category.

The importing doesn't seem to skip stuff that's already imported, and instead just imports everything again, this is annoying for when you wanna revisit lazer but if you want all those new skins in you'll have to put up with having a duplicate of every skin that was already imported, and wait ages for those few new beatmaps to be playable in lazer because you have to wait for the thousands of old ones.

If there was a system in place to detect what's new and what's old, and just skip the old, then it would be really convenient.
Of course, this isn't some game ruining thing,I mostly just play standard and revisit lazer to try out the new stuff or mess around with fun mods, so I don't NEED every new skin and beatmap, but I would love to see this done at some point before lazer's release.

**Comment by peppy** [comment, marked as answer]
it does skip anything which is already imported.

What are you seeing that suggests otherwise? Also, please provide logs after you perform an import that seems to be wrong.

**Reply by Backoffson** [reply]
As for the beatmaps, I apologize for going off my own memory instead of trying it before making this post, you're right, it does skip them.
But as for skins, it definitely duplicates them, I guess skins don't have an "ID" like beatmaps do so it makes sense.
![image](https://user-images.githubusercontent.com/37991740/177729700-a31ba95b-ec82-4a0e-9ef7-ac2774b0dfc9.png)

**Reply by peppy** [reply]
Skins still shouldn't duplicate, so you must have hit an edge case. If possible, could you please attach your `database.log` and one of the skins which is duplicating (zip the folder from your stable install and attach)?

**Reply by zanthous** [reply]
[-+Seouless+v7+-+Momoko+.zip](https://github.com/ppy/osu/files/9062105/-%2BSeouless%2Bv7%2B-%2BMomoko%2B.zip)

[database.log](https://github.com/ppy/osu/files/9062149/database.log)

here's an example, deleted a few files for filesize limit though 
![image](https://user-images.githubusercontent.com/14117191/177738932-d72fe024-deb8-4195-a3a6-ad01ae55a150.png)

the skin duplicates on every import

**Reply by peppy** [reply]
Hmm, it doesn't seem to duplicate for me.

Can you check if you potentially have a different version of that same skin in your stable skins folder?

Also can you show me the full screenshot including the folder names for that skin?

**Reply by Theighlin** [reply]
These skins get duplicated indefinitely for me

![image](https://user-images.githubusercontent.com/53081750/177738948-1aa71a1e-3b47-462a-93e4-44cf4a74767c.png)

I found the original links where i got some of these skins from because however i tried to zip them, windows wouldn't let me without changing their name, so i thought it was best to just get them as i did.

https://drive.google.com/file/d/1aD3ICHtCZcG-sdtXiVhH0ppu9xVRHKzu/view?usp=sharing
https://drive.google.com/file/d/1Q3MaHetOeWgkuxp29TM7ZHjGp_-TiWAO/view
https://archive.upsetdavid.com/osu-skins/Seysant%20III.osk
https://drive.google.com/file/d/1KKkijHh-RWUc0He7RmAJKjo30SzjS2E7/view?usp=sharing

**Reply by zanthous** [reply]
I'm pretty certain it's the correct skin, if it's the only skin in the folder I am importing, it doesn't duplicate. I'm going to investigate a bit further first

**Reply by zanthous** [reply]
It seems that it only duplicates if I am importing 10 or more beatmaps (tested 10 skins and 11 skins with a couple different skins). Probably something with batch import

here is a list of the skins I imported

![image](https://user-images.githubusercontent.com/14117191/177744687-12b4a50b-c7e0-44ac-b3d6-b7b16a02854f.png)

(some other skins with square bracket characters  '[' ']' also got duplicated)

It still says it's importing the correct amount of skins in the notification tab

**Reply by Backoffson** [reply]
Sorry for such a late reply.

[〉 Neu-R ┐.zip](https://github.com/ppy/osu/files/9062527/Neu-R.zip)
Here's one of many that duped, I didn't edit it at all.
[database.log](https://github.com/ppy/osu/files/9062533/database.log)
And the logs.

I did a full re-import, and about that screenshot I sent earlier, that was an import I quickly cancelled as I only did it to prove skins were duping, however the "DECADENCE" skin shown in it didn't dupe this time, it was skipped according to the logs.

**Reply by peppy** [reply]
Thanks, I can reproduce and will investigate further.

**Reply by peppy** [reply]
Will be resolved by #19029.

Basically, lazer will write some extra metadata to `skin.ini` in order to make sure the name displays correctly (even after export and re-import), but this was being done after a batch-import "quick" check, which didn't consider this added metadata as part of the has. So it never matched existing items.
